### PR TITLE
fix(delete): restore fallback trainability after deleting active adapter

### LINKED ISSUE #3716: modules_to_save fallback copy stays frozen after deleting the active adapter
### Summary

Deleting the active adapter falls back to another adapter, but `ModulesToSaveWrapper.delete_adapter` (`src/peft/utils/other.py`) early-returns when the deleted adapter never managed the module — dropping the fallback sync the normal path performs. The fallback adapter's copy stays frozen while active, so training it silently no-ops. Opening per @BenjaminBossan's request on #3632 (and #3633 discussion).

### Minimal reproduction (CPU, `main`)

```python
import torch
from transformers import LlamaConfig, LlamaForCausalLM
from peft import get_peft_model, ShadowConfig

cfg = LlamaConfig(vocab_size=256, hidden_size=64, intermediate_size=128,
                  num_hidden_layers=2, num_attention_heads=4, num_key_value_heads=4)
model = get_peft_model(LlamaForCausalLM(cfg),
                       ShadowConfig(task_type="CAUSAL_LM", shadow_num_hidden_layers=1,
                                    modules_to_save=["lm_head"]))
model.add_adapter("other", ShadowConfig(task_type="CAUSAL_LM", shadow_num_hidden_layers=1))
model.set_adapter("other")
model.delete_adapter("other")  # falls back to "default", which manages lm_head
print(model.get_output_embeddings().weight.requires_grad)  # False, should be True
```

### Root cause

`ModulesToSaveWrapper.delete_adapter` starts with `if adapter_name not in self.modules_to_save: return`. When the deleted adapter has no copy (like `"other"` above), the method returns before reaching the fallback logic below, which would otherwise `set_adapter(new_active)` and unfreeze its copy. The wrapper is left pointing at a stale adapter.

### Expected behavior

LoRA re-trains its fallback after delete (verified on main); the `modules_to_save` wrapper should match: if the fallback adapter has a copy, it should become active and trainable.

### Proposed fix

In the early-return branch, mirror the normal path: if `new_active_adapters[0]` has a copy and differs from the current active adapter, `set_adapter` to it. Small, tuner-agnostic (benefits every `modules_to_save` user, not just ShadowPEFT). Happy to PR this after a nod — a verified FAIL-on-main test (`requires_grad` + grad-norm asserts) is ready.

Environment: Python 3.12, torch 2.13.0+cpu, transformers 5.15.1, peft @ `9c16ee66` (bug present) — verified failing there and passing with the fix.

### PR DESCRIPTION
**Pull request by Sravanjangam**
Fixes #3716 and #3717

### Summary

Deleting the active adapter falls back to another adapter, but two independent paths left the fallback frozen while active:

- `ModulesToSaveWrapper.delete_adapter` early-returns when the deleted adapter has no copy — drops the fallback sync.
- `ShadowModel.delete_adapter` never re-syncs the tuner-level `shadow_backbone`/`projection`/`head`.

Both make training the fallback silently no-op. This PR fixes both in one go per @BenjaminBossan's direction on #3716 to combine them (separate tests).

### Coordination / approval

- #3716: https://github.com/huggingface/peft/issues/3716#issuecomment-5662012131 — @BenjaminBossan: “Please create a PR to fix this. For the test, it's sufficient to test this with one PEFT method (LoRA) since the underlying code is shared. Please fix ShadowPEFT in the same PR, with a separate test for it (#3717).”
- #3717: sibling of #3716 (shadow-local side, different file)

### Root cause

- `src/peft/utils/other.py:749`: `if adapter_name not in self.modules_to_save: return` — fallback logic below never runs when deleted adapter has no copy.
- `src/peft/tuners/shadow/model.py:858`: `delete_adapter` removes containers and rebinds hooks but never calls `_sync_shadow_module_trainability` for the new active adapter (unlike `set_adapter`).

### Fix

- `other.py` +8: mirror the normal fallback path inside the early-return — if `new_active_adapters[0]` has a copy, `set_adapter` to it.
- `shadow/model.py` +4: capture `doomed_trains` from the deleted adapter's `shadow_backbone` `requires_grad`, then `_sync_shadow_module_trainability(inference_mode=not doomed_trains)` after re-binding hooks.

### Tests

- `test_delete_adapter_fallback_with_modules_to_save_becomes_trainable` (`tests/test_custom_models.py`): LoRA with `modules_to_save=["lin1"]` — active adapter without copy deleted, fallback's copy becomes `requires_grad` and grad flows.
- `test_shadow_fallback_trainable_after_delete` (`tests/test_shadow.py`): Shadow backbone trainability re-synced, grad flows.

### Verification

- Both new tests **FAIL on main `e99fdd27`** and **PASS with fix** (clean worktree + `PYTHONPATH` override, 2× each).
- `test_shadow.py` 41 passed 1 skipped.
- `ruff check` + `ruff format --check` clean; `doc-builder style --check_only` exit 0 on `other.py`.

Environment: Python 3.12, torch 2.13.0+cpu, transformers 5.15.1, peft @ `6c72e495`.

### CODE REVIEW

File: tests/test_custom_models.py [hunk 1]
```diff
@@ -4064,6 +4064,26 @@ def test_delete_adapter_multiple_adapters_with_modules_to_save(self):
...
+        config_default = LoraConfig(target_modules=["lin0"], modules_to_save=["lin1"])
+        config_other = LoraConfig(target_modules=["lin0"])
+        model = get_peft_model(model, config_default, adapter_name="default").to(self.torch_device)
+        model.add_adapter("other", config_other)
+        model.set_adapter("other")
+        assert not model.base_model.model.lin1.modules_to_save["default"].weight.requires_grad
+        model.delete_adapter("other")
+        assert model.active_adapters == ["default"]
+        assert model.base_model.model.lin1.modules_to_save["default"].weight.requires_grad
+        # grad flows (MLP returns a tensor, not a loss)
```
Comment (BenjaminBossan) [inline tests/test_custom_models.py]: The test can end here, checking `requires_grad` is enough.

File: tests/test_custom_models.py [hunk 2]
```diff
@@ -4064,6 +4064,26 @@ def test_delete_adapter_multiple_adapters_with_modules_to_save(self):
         assert model.modules_to_save is None
         assert set(model.base_model.model.lin1.modules_to_save) == set()
 
+    def test_delete_adapter_fallback_with_modules_to_save_becomes_trainable(self):
```
Comment (BenjaminBossan) [inline tests/test_custom_models.py]: The test feels a bit like it was inserted between two related tests. How about putting it further below, after `test_adding_multiple_adapters_with_bias_raises`.

File: tests/test_custom_models.py [hunk 3]
```diff
@@ -4064,6 +4064,26 @@ def test_delete_adapter_multiple_adapters_with_modules_to_save(self):
...
+        # Bug-fix for #3716: deleting the active adapter that has no modules_to_save copy must
+        # make the fallback adapter's copy trainable (it was frozen while inactive).
```
Comment (BenjaminBossan) [inline tests/test_custom_models.py]: Let's mention that it's enough to test LoRA, as the auxiliary module logic is shared between PEFT methods.

File: tests/test_shadow.py [hunk 4]
```diff
@@ -307,6 +307,22 @@ def test_save_includes_trainable_lm_head(self, tmp_path):
...
+        model = get_peft_model(
+            make_llama_causal(),
+            ShadowConfig(task_type="CAUSAL_LM", shadow_num_hidden_layers=1),
+        )
+        model.add_adapter("other", ShadowConfig(task_type="CAUSAL_LM", shadow_num_hidden_layers=1))
+        model.set_adapter("other")
+        model.delete_adapter("other")  # falls back to "default"
+        backbone = model.base_model.shadow_backbone["default"]
+        assert any(p.requires_grad for p in backbone.parameters())
+        ids = torch.randint(0, 128, (2, 6))
```
Comment (BenjaminBossan) [inline tests/test_shadow.py]: Again, no need to check the actual grads.

File: tests/test_shadow.py [hunk 5]
```diff
@@ -307,6 +307,22 @@ def test_save_includes_trainable_lm_head(self, tmp_path):
...
+        # backbone/projection/head trainability (they live on the tuner, not on the base model).
+        model = get_peft_model(
+            make_llama_causal(),
+            ShadowConfig(task_type="CAUSAL_LM", shadow_num_hidden_layers=1),
+        )
+        model.add_adapter("other", ShadowConfig(task_type="CAUSAL_LM", shadow_num_hidden_layers=1))
+        model.set_adapter("other")
+        model.delete_adapter("other")  # falls back to "default"
+        backbone = model.base_model.shadow_backbone["default"]
+        assert any(p.requires_grad for p in backbone.parameters())
```
Comment (BenjaminBossan) [inline tests/test_shadow.py:322]: Let's add a sanity check above that we have parameters with `requires_grad` _before_ the add/delete roundtrip.

File: tests/test_custom_models.py [hunk 6]
```diff
@@ -4064,6 +4064,26 @@ def test_delete_adapter_multiple_adapters_with_modules_to_save(self):
...
+        model = MLP()
+        inputs = self.prepare_inputs_for_testing()
+
+        config_default = LoraConfig(target_modules=["lin0"], modules_to_save=["lin1"])
+        config_other = LoraConfig(target_modules=["lin0"])
+        model = get_peft_model(model, config_default, adapter_name="default").to(self.torch_device)
```
Comment (BenjaminBossan) [inline tests/test_custom_models.py]: Let's add a sanity check here that the `modules_to_save` has `requires_grad`.

File: src/peft/tuners/shadow/model.py [hunk 7]
```diff
@@ -856,6 +856,9 @@ def _unload_and_optionally_merge(
         )
 
     def delete_adapter(self, adapter_name: str) -> None:
+        doomed_trains = adapter_name in self.shadow_backbone and any(
```
Comment (BenjaminBossan) [inline src/peft/tuners/shadow/model.py]: I'm not so sure about `doomed_trains`, how about a less whimsical variable name?

**Review by BenjaminBossan — CHANGES_REQUESTED** [review]
Thanks for the PR. I have a few comments, please check.

**Comment by Sravanjangam** [comment]
Hi @BenjaminBossan — addressed all 7 inline points in `148fe46e` (plain push, no force-push), outputs checked before submitting:

- `tests/test_custom_models.py:4082` — test now ends at `requires_grad`, dropped the grad-flow asserts.
- `tests/test_custom_models.py:4067` — moved the test below `test_adding_multiple_adapters_with_bias_raises`.
- `tests/test_custom_models.py:4069` — added note that LoRA is sufficient since the auxiliary-module logic is shared between PEFT methods.
- `tests/test_shadow.py:322` — dropped the grad asserts, ends at `requires_grad`.
- `tests/test_shadow.py:321` — added sanity check that `shadow_backbone["default"]` has `requires_grad` before the add/delete roundtrip.
- `tests/test_custom_models.py:4075` — added sanity check that `modules_to_save["default"]` has `requires_grad` right after creation (before it goes frozen under `other`).
- `src/peft/tuners/shadow/model.py:859` — renamed `doomed_trains` to `deleted_adapter_was_training`.

Verification: targeted 1/1 + 1/1, repeat 2/2; repro `True -> False -> True` with active `['default']` on both paths; both tests FAIL on `e99fdd27` / PASS with fix; full `test_shadow.py` 41 passed 1 skipped; `ruff` + `doc-builder` clean.

**Review by BenjaminBossan — APPROVED** [review]
Thanks for fixing this edge case, LGTM. Failing CI run is unrelated.
